### ISSUE TITLE
Unexpected behaviour with past_key_values in GPT2

### ISSUE THREAD

**Issue by YoadTew** [huggingface/transformers#13849]
## Environment info

- `transformers` version: 4.11.2
- Platform: Ubuntu 20.04
- Python version: 3.8.10
- PyTorch version (GPU?): 1.9 GPU
- Tensorflow version (GPU?):
- Using GPU in script?: rtx 3070
- Using distributed or parallel set-up in script?: no

### Who can help
@patrickvonplaten, @LysandreJik

## Information

Model I am using (Bert, XLNet ...): GPT2LMHeadModel,  gpt2-medium

When using GPT2LMHeadModel with past_key_values as an input, the output logits are different than expected.
I expected that when I run inference with the context ''Image of a" and when I run inference with the context "a" and past_key_values of previous run with the context "Image of", the output logits will be the same, but they are slightly different. I added code below.
Is this a bug, or am I missing something?

## To reproduce

Run the code:

```
from transformers.models.gpt2 import GPT2LMHeadModel, GPT2Tokenizer
import numpy as np
import torch

seed = 0
device = "cuda" if torch.cuda.is_available() else "cpu"
# set Random seed
torch.manual_seed(seed)
np.random.seed(seed)

lm_tokenizer = GPT2Tokenizer.from_pretrained('gpt2-medium')
lm_model = GPT2LMHeadModel.from_pretrained('gpt2-medium', output_hidden_states=True)
lm_model.to(device)
lm_model.eval()

raw_text = 'Image of a'
tokenized_cond_text = lm_tokenizer.encode(lm_tokenizer.bos_token + raw_text)

context_t = torch.tensor(tokenized_cond_text, device=device, dtype=torch.long)
while len(context_t.shape) < 2:
    context_t = context_t.unsqueeze(0)
output_so_far = context_t

past = lm_model(output_so_far[:, :-1])["past_key_values"]
last = output_so_far[:, -1:]

logits_past = lm_model(last, past_key_values=past)['logits']
logits_normal = lm_model(output_so_far)['logits'][:, -1:, :]

print(logits_normal.mean().cpu().item(), logits_past.mean().cpu().item())
```
My output:
`-59.50511169433594 -59.50926971435547`

## Expected behavior
I expected for the logits to be the same (and therfore have the same mean). Am I missing something?

**Comment by YoadTew** [comment]
@patrickvonplaten Oh really, I didn't know it. So the masked self attention is not perfect and it's the reason for the difference.

The reason I noticed it is that the slight difference resulted in a slightly diffrent text generation. Thanks for the response.

**Comment by patrickvonplaten** [comment]
I'm afraid you probably have to edit the Hugging Face source code at the moment...out of curiosity - what is the use case that requires `use_cache=True` to give exactly the same outputs as `use_cache=False`?
